python/src/com/jetbrains/python/psi/types/PyImportedModuleType.java

```diff
@@ -43,20 +37,33 @@ public PyImportedModuleType(@NotNull PyImportedModule importedModule) {
     final PsiElement resolved = myImportedModule.resolve();
     if (resolved != null) {
       final PsiFile containingFile = location != null ? location.getContainingFile() : null;
-      List<PsiElement> elements = Collections.singletonList(ResolveImportUtil.resolveChild(resolved, name, containingFile, false, true,
-                                                                                           false));
+      PsiElement resolvedChild = ResolveImportUtil.resolveChild(resolved, name, containingFile, false, true,
+                                                                false);
       final PyImportElement importElement = myImportedModule.getImportElement();
       final PyFile resolvedFile = PyUtil.as(resolved, PyFile.class);
       if (location != null && importElement != null && PyUtil.inSameFile(location, importElement) &&
           ResolveImportUtil.getPointInImport(location) == PointInImport.NONE && resolved instanceof PsiFileSystemItem &&
           (resolvedFile == null || !PyUtil.isPackage(resolvedFile) || resolvedFile.getElementNamed(name) == null)) {
-        final List<PsiElement> importedSubmodules = PyModuleType.collectImportedSubmodules((PsiFileSystemItem)resolved, location);
-        if (importedSubmodules != null) {
-          final Set<PsiElement> imported = Sets.newHashSet(importedSubmodules);
-          elements = ContainerUtil.filter(elements, element -> imported.contains(element));
+
+        ResolveResultList res = new ResolveResultList();
+        boolean isPackage =
+          PyModuleType.processImportedSubmodules((PsiFileSystemItem)resolved, location, (nameDefiner, importedSubmodule) -> {
+            if (importedSubmodule.equals(resolvedChild)) {
+              res.add(new ImportedResolveResult(resolvedChild, RatedResolveResult.RATE_NORMAL, nameDefiner));
+              return false;
+            }
+            return true;
+          });
+
+        if (isPackage) {
+          return res;
```

Comment:
I'm not sure what is the key change, `isPackage` flag? What is the smallest part of code to comment or revert to make the test failed?

Reply:
This is just refactoring. Changing returning a list or null to a processor. This is because it was unclear what was going on PyImportedModuleType.resolveMember. It was looking for a single element by filtering a list against a set containing a single item.  Using a processor makes it clear we're looking for a single element and is also more efficient since we  can stop once we find it. 

The only key change is wrapping the return values of PyImportedModuleType.resolveMember as `ImportedResolveResult`

Reply:
The isPackage flag was because the return value of the function was treated differently if it returned emptyList and null.  I think I can make it cleaner by making it return `void` and adding the package check seperately

Reply:
Thank you, it's clear to me now